db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/NumericTypeCompiler.java

```diff
@@ -364,6 +364,53 @@ else if (typeId.isBigIntTypeId() &&
 			}
 			else if (typeId.isDecimalTypeId())
 			{
+				/* DB-9425
+				 * Make sure we still have enough digits for the integral part.
+				 * Reduce scale when necessary, or leave it when not possible.
+				 */
+				{
+					int lprec = leftType.getPrecision();
+					int lscale = leftType.getScale();
+					int rprec = rightType.getPrecision();
+					int rscale = rightType.getScale();
+					int integralNumDigits;
+
+					switch (operator) {
+						case TypeCompiler.SUM_OP:
+						case TypeCompiler.MINUS_OP:
+							boolean addOne = (lprec != maxPrecision && rprec != maxPrecision);
+							integralNumDigits = Math.max(lprec - lscale, rprec - rscale) + (addOne ? 1 : 0);
+							if (integralNumDigits <= precision && integralNumDigits + scale > precision) {
+								scale = precision - integralNumDigits;
+							}
+							break;
+						case TypeCompiler.TIMES_OP: {
+							int fullScale = lscale + rscale;
+							int fullPrec = lprec + rprec + 1;
+							integralNumDigits = fullPrec - fullScale;
+							if (integralNumDigits <= precision &&
+								integralNumDigits + scale > precision && scale > MIN_DECIMAL_DIVIDE_SCALE &&
+							    precision - integralNumDigits >= MIN_DECIMAL_DIVIDE_SCALE) {
+								scale = precision - integralNumDigits;
+							}
+							break;
+						}
+						case TypeCompiler.DIVIDE_OP: {
+							if (rightTypeId.isDecimalTypeId()) {
+								integralNumDigits = lprec - lscale + rprec;
```

Comment:
should it be rscale instead of rprec?

Reply:
It seems that for the division precision and scale computation, the logic in getScale() and getPrecision() are not consistent.
In getPrecision():
```
			val = Math.min(maxPrecision,
						   this.getScale(operator, leftType, rightType) + lprec - lscale + rprec);
```
In getScale():
```
				val = Math.max(maxPrecision - lprec + lscale - rscale,
						MIN_DECIMAL_DIVIDE_SCALE);
```
I think the formula in getPrecision should be changed to 
```
			val = Math.min(maxPrecision,
						   this.getScale(operator, leftType, rightType) + lprec - lscale + rscale);
```
to be consistent with getScale().

Reply:
This is the mystery part and I think it was done like this on purpose. The idea seems to be always use `maxPrecision` for division and leave all free digits to scale. Let's take a closer look:

The way we get scale for division, is to calculate the number of digits needed for the integral part (`lprec - lscale + rscale`), and subtract it from `maxPrecision`. When getting precision, we add to scale a term `lprec - lscale + rprec` that is always GE than the integral part, so it will be always `maxPrecision` in the end. This change was done in implementing Decimal 38. The logic of getting scale for division was not like this, and we can see that from the old comment.

Changing the logic back to the old comment or the SQL server way causes quite many tests to fail, including TPCH ITs. I'm examining the failing tests one by one.

Reply:
I finally get the root cause of test failures: SQL Server's MIN_DECIMAL_DIVIDE_SCALE = 6, but we set it to 4, and I was using 4 in multiplication case, too. That causes TPC-H queries return wrong results.

I fixed this part and also revert the logic of getPrecision() for division according to the old comment. Now only a few tests need to be adapted.